Change `binary_asm_labels` to only fire on x86 and x86_64

tgross35 opened the pull request:
In <https://github.com/rust-lang/rust/pull/126922>, the `binary_asm_labels` lint was added which flags labels such as `0:` and `1:`. Before that change, LLVM was giving a confusing error on x86/x86_64 because of an incorrect interpretation.

However, targets other than x86 and x86_64 never had the error message and have not been a problem. This means that the lint was causing code that previously worked to start failing (e.g. `compiler_builtins`), rather than only providing a more clear messages where there has always been an error.

Adjust the lint to only fire on x86 and x86_64 assembly to avoid this regression.

Also update the help message.

Fixes: https://github.com/rust-lang/rust/issues/127821

<!-- homu-ignore:start -->
<!-- homu-ignore:end -->

tgross35 committed:
Change `binary_asm_labels` to only fire on x86 and x86_64

In <https://github.com/rust-lang/rust/pull/126922>, the
`binary_asm_labels` lint was added which flags labels such as `0:` and
`1:`. Before that change, LLVM was giving a confusing error on
x86/x86_64 because of an incorrect interpretation.

However, targets other than x86 and x86_64 never had the error message
and have not been a problem. This means that the lint was causing code
that previously worked to start failing (e.g. `compiler_builtins`),
rather than only providing a more clear messages where there has always
been an error.

Adjust the lint to only fire on x86 and x86_64 assembly to avoid this
regression.

tgross35 committed:
Update the `binary_asm_label` message

The link pointed to a closed issue. Create a new one and point the link
to it.

Also add a help message to hint what change the user could make.

Fixes: https://github.com/rust-lang/rust/issues/127821

tgross35 commented on compiler/rustc_lint/src/lints.rs:
```diff
@@ -2066,6 +2066,7 @@ pub enum InvalidAsmLabel {
         missing_precise_span: bool,
     },
     #[diag(lint_invalid_asm_label_binary)]
+    #[help]
```
What do I need to do to make this propagate to https://github.com/rust-lang/rust/blob/5753b3067662e17a69b54b9418dbc37b73769a84/tests/ui/asm/binary_asm_labels.stderr#L1-L8?

Urgau replied:
It should work. The test is gated on `only-x86_64`, so if you weren't on x86_64 it wouldn't be blessed.

tgross35 replied:
Ah, maybe I should try running it on an actual x86 machine

Urgau commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -403,8 +403,9 @@ lint_inner_macro_attribute_unstable = inner macro attributes are unstable
     .label = use a different label that doesn't start with `0` or `1`
-    .note = an LLVM bug makes these labels ambiguous with a binary literal number
-    .note = see <https://bugs.llvm.org/show_bug.cgi?id=36144> for more information
+    .help = Start numbering with `2` instead
```
Diagnostic messages should start with a lower case letter.
```suggestion
    .help = start numbering with `2` instead
```

Urgau commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -403,8 +403,9 @@ lint_inner_macro_attribute_unstable = inner macro attributes are unstable
-    .note = see <https://bugs.llvm.org/show_bug.cgi?id=36144> for more information
+    .help = Start numbering with `2` instead
+    .note = an LLVM bug makes these labels ambiguous with a binary literal number on x86
+    .note = see <https://github.com/llvm/llvm-project/issues/99547> for more information
```
Pre-existing, but I don't think this works like this; this line should use a different identifier and be referenced separately in the `derive(LintDiagnostic)` (i.e. `#[note(lint_issue)]` or whatever).

tgross35 replied:
Oh, that's probably why my `.note` wasn't showing up either. Thanks, I'll try it

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9998078923/job/27636173430) [(plain)](https://github.com/rust-lang/rust/commit/cc4b217d13fd6ba1e3e678724706380548ee5e41/checks/27636173430/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:3aacb9c90579defe09351ac5e8ee504359f8054da6326ff19038f1b7c90e3cb2aafe33685c6d9b76ee8d2ccbd187ca80c46ab5380485abdd8c0ce7d69cd8d8fd:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

---- [ui] tests/ui/asm/binary_asm_labels.rs stdout ----
diff of stderr:

4 LL |         asm!("0: jmp 0b");
5    |               ^ use a different label that doesn't start with `0` or `1`
6    |
-    = note: an LLVM bug makes these labels ambiguous with a binary literal number
+    = help: Start numbering with `2` instead
+    = note: an LLVM bug makes these labels ambiguous with a binary literal number on x86
8    = note: `#[deny(binary_asm_labels)]` on by default
10 error: avoid using labels containing only the digits `0` and `1` in inline assembly


13 LL |         asm!("1: jmp 1b");
14    |               ^ use a different label that doesn't start with `0` or `1`
15    |
-    = note: an LLVM bug makes these labels ambiguous with a binary literal number
+    = help: Start numbering with `2` instead
+    = note: an LLVM bug makes these labels ambiguous with a binary literal number on x86
18 error: avoid using labels containing only the digits `0` and `1` in inline assembly
19   --> $DIR/binary_asm_labels.rs:13:15


21 LL |         asm!("10: jmp 10b");
22    |               ^^ use a different label that doesn't start with `0` or `1`
23    |
-    = note: an LLVM bug makes these labels ambiguous with a binary literal number
+    = help: Start numbering with `2` instead
+    = note: an LLVM bug makes these labels ambiguous with a binary literal number on x86
26 error: avoid using labels containing only the digits `0` and `1` in inline assembly
27   --> $DIR/binary_asm_labels.rs:14:15


29 LL |         asm!("01: jmp 01b");
30    |               ^^ use a different label that doesn't start with `0` or `1`
31    |
-    = note: an LLVM bug makes these labels ambiguous with a binary literal number
+    = help: Start numbering with `2` instead
+    = note: an LLVM bug makes these labels ambiguous with a binary literal number on x86
34 error: avoid using labels containing only the digits `0` and `1` in inline assembly
35   --> $DIR/binary_asm_labels.rs:15:15


37 LL |         asm!("1001101: jmp 1001101b");
38    |               ^^^^^^^ use a different label that doesn't start with `0` or `1`
39    |
-    = note: an LLVM bug makes these labels ambiguous with a binary literal number
+    = help: Start numbering with `2` instead
+    = note: an LLVM bug makes these labels ambiguous with a binary literal number on x86
42 error: aborting due to 5 previous errors
43 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/binary_asm_labels/binary_asm_labels.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args asm/binary_asm_labels.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/asm/binary_asm_labels.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/binary_asm_labels" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/binary_asm_labels/auxiliary"
--- stderr -------------------------------
error: avoid using labels containing only the digits `0` and `1` in inline assembly
##[error]  --> /checkout/tests/ui/asm/binary_asm_labels.rs:11:15
   |
   |
LL |         asm!("0: jmp 0b"); //~ ERROR avoid using labels containing only the digits
   |               ^ use a different label that doesn't start with `0` or `1`
   = help: Start numbering with `2` instead
   = help: Start numbering with `2` instead
   = note: an LLVM bug makes these labels ambiguous with a binary literal number on x86
   = note: `#[deny(binary_asm_labels)]` on by default
error: avoid using labels containing only the digits `0` and `1` in inline assembly
##[error]  --> /checkout/tests/ui/asm/binary_asm_labels.rs:12:15
   |
   |
LL |         asm!("1: jmp 1b"); //~ ERROR avoid using labels containing only the digits
   |               ^ use a different label that doesn't start with `0` or `1`
   = help: Start numbering with `2` instead
   = help: Start numbering with `2` instead
   = note: an LLVM bug makes these labels ambiguous with a binary literal number on x86
error: avoid using labels containing only the digits `0` and `1` in inline assembly
##[error]  --> /checkout/tests/ui/asm/binary_asm_labels.rs:13:15
   |
   |
LL |         asm!("10: jmp 10b"); //~ ERROR avoid using labels containing only the digits
   |               ^^ use a different label that doesn't start with `0` or `1`
   = help: Start numbering with `2` instead
   = help: Start numbering with `2` instead
   = note: an LLVM bug makes these labels ambiguous with a binary literal number on x86
error: avoid using labels containing only the digits `0` and `1` in inline assembly
##[error]  --> /checkout/tests/ui/asm/binary_asm_labels.rs:14:15
   |
   |
LL |         asm!("01: jmp 01b"); //~ ERROR avoid using labels containing only the digits
   |               ^^ use a different label that doesn't start with `0` or `1`
   = help: Start numbering with `2` instead
   = help: Start numbering with `2` instead
   = note: an LLVM bug makes these labels ambiguous with a binary literal number on x86
error: avoid using labels containing only the digits `0` and `1` in inline assembly
##[error]  --> /checkout/tests/ui/asm/binary_asm_labels.rs:15:15
   |
   |
LL |         asm!("1001101: jmp 1001101b"); //~ ERROR avoid using labels containing only the digits
   |               ^^^^^^^ use a different label that doesn't start with `0` or `1`
   = help: Start numbering with `2` instead
   = help: Start numbering with `2` instead
   = note: an LLVM bug makes these labels ambiguous with a binary literal number on x86
error: aborting due to 5 previous errors
------------------------------------------


```

</details>

tgross35 commented:
Tested locally on both x86 and aarch64.

r? @Urgau

estebank commented:
Ideally, the lint would not gate on the platform being compiled, but rather gated on whether the code *could* be run from x86. Namely, if the `asm` is behind a `cfg` flag that limits the code to anything other than x86. That being said, I believe that for now this is a reasonable compromise.

@bors r+

bors commented:
:pushpin: Commit 8410348b1c0981b60500ec1f571f6ed8fc075fc8 has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank 8410348b1c0981b60500ec1f571f6ed8fc075fc8 -->
<!-- homu: {"type":"Approved","sha":"8410348b1c0981b60500ec1f571f6ed8fc075fc8","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Urgau commented on compiler/rustc_lint/messages.ftl:
```diff
@@ -403,8 +403,9 @@ lint_inner_macro_attribute_unstable = inner macro attributes are unstable
-    .note = an LLVM bug makes these labels ambiguous with a binary literal number
-    .note = see <https://bugs.llvm.org/show_bug.cgi?id=36144> for more information
+    .help = start numbering with `2` instead
+    .note1 = an LLVM bug makes these labels ambiguous with a binary literal number on x86
+    .note2 = see <https://github.com/llvm/llvm-project/issues/99547> for more information
```
nit: it isn't necessary for both of them to use a different identifier, only of them needs to, the other one can stay as always (`.note = ...` and `#[note]`)

tgross35 replied:
I just mirrored a few lines up in https://github.com/rust-lang/rust/pull/127935/files#diff-4a2362ab93352b636e03b023fecbe61e96c3a07f1be1393d67bbd87d87374f50L2060-L2063

Seems like we should have a way to detect unused `.note`s somehow, since the second wasn't being used until this change.

Urgau replied:
> Seems like we should have a way to detect unused labels, since the second .note wasn't being used until this change.

We should indeed.

Anyway, no need to change anything. Just wanted you to be aware of this.

tgross35 replied:
Looks like it's at least been considered https://github.com/rust-lang/rust/issues/110754

estebank commented:
> Does that kind of machinery exist? I thought if it was behind a cfg then it doesn't even know the asm is there.

We have a very basic [collection of items that got cfg'd out for the purposes of telling users that an item that wasn't found is behind a feature flag](https://github.com/rust-lang/rust/blob/8eb51852a87dd2640fbd95019ff1124b38ae8b78/compiler/rustc_resolve/src/diagnostics.rs#L2502-L2540). [We collect those during expansion](https://github.com/rust-lang/rust/blob/a23917cfd07b309f32603c6acb81b63aad8504e7/compiler/rustc_expand/src/expand.rs#L1964-L1968), which would make my request harder to accomplish, as you would need to implement the lint to be able to work with an `ast::ExprKind::InlineAsm` (as opposed to the current `hir::ExprKind::InlineAsm`) but there's precedent.

tgross35 commented:
For this specific lint, it seems like effort is probably better spent fixing LLVM than attempting to improve the ergonomics much more. But being able to leverage that for other inline asm sounds amazing, since it is so platform dependent and easy to make mistakes, we should be able to at least perform Rust's validation on it.

I'll update https://github.com/rust-lang/rust/issues/127938 to reflect that if it sounds accurate.

tgross35 commented:
@bors r- failed rollup https://github.com/rust-lang/rust/pull/127941#issuecomment-2238208707

tgross35 committed:
Update the `binary_asm_label` documentation

Disable a test that now only passes on x86 and make the link point to
the new (open) LLVM bug.

tgross35 commented:
I didn't update the docs so it tried to run the assembly example on all platforms. Added a commit to fix this, looking at other tests I guess the best thing we can do is `ignore` the doctest and update the output manually. `lint-docs` tests pass locally on my aarch64 machine now.

I'll let somebody double check the changes (@estebank), @rustbot review

Urgau commented:
Changes looks good.

@bors r=estebank,Urgau

bors commented:
:pushpin: Commit 5686720a5706afa5628e1ff00cb6622eb1399d35 has been approved by `estebank,Urgau`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank,Urgau 5686720a5706afa5628e1ff00cb6622eb1399d35 -->
<!-- homu: {"type":"Approved","sha":"5686720a5706afa5628e1ff00cb6622eb1399d35","approver":"estebank,Urgau","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
